### PR TITLE
feat: add runtime recording feature

### PR DESCRIPTION
**Pull request by hsgwa**
This PR adds runtime recording feature.

Following PR are needed to enable.

- https://github.com/tier4/CARET_analyze/pull/190
- https://github.com/tier4/ros2caret/pull/48

### CODE REVIEW

File: CARET_trace/include/caret_trace/trace_node.hpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+#include <memory>
+#include <string>
+#include <utility>
+
+enum class TRACE_STATUS {
+  UNINITIALIZED,
+  WAIT,
+  PREPARE,
+  RECORD,
+};
```
[the remark this answers is not in the document]
Comment (hsgwa) [reply CARET_trace/include/caret_trace/trace_node.hpp:34]: Umm..
Then, hou about following?
- `WAITING` state
- `SETUP` state
- `RECORDING` state

File: CARET_trace/src/ros_trace_points.cpp [hunk 2]
```diff
@@ -83,11 +185,17 @@ void ros_trace_rcl_node_init(
 
   controller.add_node(node_handle, node_ns_and_name);
 
-  using functionT = void (*)(const void *, const void *, const char *, const char *);
+  auto now = clock.now();
 
-  if (controller.is_allowed_node(node_handle)) {
+  bool pending = data_container.store_rcl_node_init(
```
[the remark this answers is not in the document]
Comment (hsgwa) [reply CARET_trace/src/ros_trace_points.cpp:205]: `rcl_node_init.store` is not `std::atomic::store`.
It's defined here: (recordable_data.hpp)
https://github.com/tier4/CARET_trace/pull/68/files#diff-e765e792a6be291e632242f9dfd47e0598f687d24ebea5282627e10e80f990a1R141

[the remark this answers is not in the document]
Comment (hsgwa) [reply CARET_trace/src/ros_trace_points.cpp:205]: @keita1523

File: CARET_trace/include/caret_trace/recordable_data.hpp [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    }
+    pending_set_.clear();
+  }
+
+  /// @brief merge_pending_data when iterator reached to the end.
+  /// @return true : Succeed to merge.
+  /// @return false : No data were merged.
+  bool try_merge_pending_data()
+  {
+    if (is_end__) {
```
Comment (keita1523) [inline CARET_trace/include/caret_trace/recordable_data.hpp:243]: Please add comments what is end

File: CARET_trace/src/context.cpp [hunk 4]
```diff
@@ -14,18 +14,77 @@
...
+{
+  if (is_node_assigned()) {
+    return node_->is_recording_allowed();
+  }
+
+  // NOTE: Prohibited to suppress DISCARDED.
+  return false;
+}
+
+bool Context::is_recording_allowed_init() const
```
Comment (keita1523) [inline CARET_trace/src/context.cpp:78]: Is this function unused? Please add comments

File: CARET_trace/src/lttng_session.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#include <chrono>
+#include <iostream>
+#include <memory>
+#include <mutex>
+
+bool LttngSessionImpl::is_session_running() const
+{
+  std::lock_guard<std::mutex> lock(mtx_);
+
+  std::string command = "lttng list | grep -q \'\\[active\\]\'";
```
Comment (keita1523) [inline CARET_trace/src/lttng_session.cpp:28]: Please add comments.

File: CARET_trace/src/ros_trace_points.cpp [hunk 6]
```diff
@@ -83,20 +203,10 @@ void ros_trace_rcl_node_init(
...
 
-#ifdef DEBUG_OUTPUT
-    std::cerr << "rcl_node_init," <<
-      node_handle << "," <<
-      rmw_handle << "," <<
-      node_name << "," <<
-      node_namespace << std::endl;
-#endif
-  }
+    record(node_handle, rmw_handle, node_name, node_namespace, now);
```
Comment (keita1523) [inline CARET_trace/src/ros_trace_points.cpp:209]: Wrong indent

File: CARET_trace/src/ros_trace_points.cpp [hunk 7]
```diff
@@ -127,99 +244,191 @@ void ros_trace_rcl_subscription_init(
...
+
+  check_and_run_trace_node();
+
+  data_container.store_rcl_subscription_init(
+    subscription_handle, node_handle, rmw_subscription_handle, topic_name,
+    reinterpret_cast<size_t>(queue_depth), now);
+
+    record(
+      subscription_handle, node_handle, rmw_subscription_handle, topic_name,
+      queue_depth, now);
```
Comment (keita1523) [inline CARET_trace/src/ros_trace_points.cpp:265]: Wrong indent

File: CARET_trace/src/ros_trace_points.cpp [hunk 8]
```diff
@@ -127,99 +244,191 @@ void ros_trace_rcl_subscription_init(
...
+  controller.add_subscription(subscription_handle, subscription);
+
+  if (!data_container.is_assigned_rclcpp_subscription_init()) {
+    data_container.assign_rclcpp_subscription_init(record);
   }
+
+  data_container.store_rclcpp_subscription_init(
+    subscription_handle, subscription, now);
+
+    record(subscription_handle, subscription, now);
```
Comment (keita1523) [inline CARET_trace/src/ros_trace_points.cpp:307]: Wrong indent

File: CARET_trace/src/ros_trace_points.cpp [hunk 9]
```diff
@@ -127,99 +244,191 @@ void ros_trace_rcl_subscription_init(
...
+  controller.add_subscription_callback(subscription, callback);
+
+  if (!data_container.is_assigned_rclcpp_subscription_callback_added()) {
+    data_container.assign_rclcpp_subscription_callback_added(record);
   }
+
+  data_container.store_rclcpp_subscription_callback_added(
+    subscription, callback, now);
+
+    record(subscription, callback, now);
```
Comment (keita1523) [inline CARET_trace/src/ros_trace_points.cpp:347]: Wrong indent

File: CARET_trace/src/ros_trace_points.cpp [hunk 10]
```diff
@@ -127,99 +244,191 @@ void ros_trace_rcl_subscription_init(
...
+  check_and_run_trace_node();
+
+  controller.add_timer_callback(timer_handle, callback);
+
+  if (!data_container.is_assigned_rclcpp_timer_callback_added()) {
+    data_container.assign_rclcpp_timer_callback_added(record);
   }
+
+  data_container.store_rclcpp_timer_callback_added(timer_handle, callback, now);
+    record(timer_handle, callback, now);
```
Comment (keita1523) [inline CARET_trace/src/ros_trace_points.cpp:382]: Wrong indent

File: CARET_trace/src/ros_trace_points.cpp [hunk 11]
```diff
@@ -127,99 +244,191 @@ void ros_trace_rcl_subscription_init(
...
+  check_and_run_trace_node();
+
+  controller.add_timer_handle(node_handle, timer_handle);
+
+  if (!data_container.is_assigned_rclcpp_timer_link_node()) {
+    data_container.assign_rclcpp_timer_link_node(record);
   }
+
+  data_container.store_rclcpp_timer_link_node(timer_handle, node_handle, now);
+    record(timer_handle, node_handle, now);
```
Comment (keita1523) [inline CARET_trace/src/ros_trace_points.cpp:417]: Wrong indent

File: CARET_trace/src/ros_trace_points.cpp [hunk 12]
```diff
@@ -334,36 +553,58 @@ void ros_trace_rclcpp_intra_publish(
...
+    tracepoint(TRACEPOINT_PROVIDER, rcl_timer_init, timer_handle, period, init_time);
+  };
+
+  auto now = clock.now();
+  if (!data_container.is_assigned_rcl_timer_init()) {
+    data_container.assign_rcl_timer_init(record);
+  }
+
+  data_container.store_rcl_timer_init(timer_handle, period, now);
+    record(timer_handle, period, now);
```
Comment (keita1523) [inline CARET_trace/src/ros_trace_points.cpp:575]: Wrong indent

File: CARET_trace/src/ros_trace_points.cpp [hunk 13]
```diff
@@ -334,36 +553,58 @@ void ros_trace_rclcpp_intra_publish(
...
+    tracepoint(TRACEPOINT_PROVIDER, rcl_init, context_handle, init_time);
+  };
 
+  auto now = clock.now();
+  if (!data_container.is_assigned_rcl_init()) {
+    data_container.assign_rcl_init(record);
+  }
+
+  data_container.store_rcl_init(context_handle, now);
+    record(context_handle, now);
```
Comment (keita1523) [inline CARET_trace/src/ros_trace_points.cpp:601]: Wrong Indent

File: CARET_trace/src/ros_trace_points.cpp [hunk 14]
```diff
@@ -374,41 +615,71 @@ void ros_trace_rcl_publisher_init(
...
+    queue_depth,
+    now);
+
+    record(
+      publisher_handle,
+      node_handle,
+      rmw_publisher_handle,
+      topic_name,
+      queue_depth,
+      now);
```
Comment (keita1523) [inline CARET_trace/src/ros_trace_points.cpp:659]: Wrong Indent

File: CARET_trace/src/ros_trace_points.cpp [hunk 15]
```diff
@@ -417,128 +688,225 @@ void ros_trace_rcl_publish(
...
+  if (!data_container.is_assigned_rcl_service_init()) {
+    data_container.assign_rcl_service_init(record);
+  }
+
+  check_and_run_trace_node();
+
+  data_container.store_rcl_service_init(
+    service_handle, node_handle, rmw_service_handle,
+    service_name, now);
+    record(service_handle, node_handle, rmw_service_handle, service_name, now);
```
Comment (keita1523) [inline CARET_trace/src/ros_trace_points.cpp:727]: Wrong Indent

File: CARET_trace/src/trace_node.cpp [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+{
+  // NOTE: Trace points for measurement are recorded without storing in memory.
+  // The trace point for measurement should be forbidden in PREPARE state to suppress DISCARDED.
+  // No strict control is required, so no mutex is needed.
+  return status_ == TRACE_STATUS::RECORD;
+}
+
+bool TraceNode::is_recording_allowed_init() const
+{
+  return true;
```
Comment (keita1523) [inline CARET_trace/src/trace_node.cpp:147]: Please add comments.

**Review by keita1523 — COMMENTED** [review]
Minor request

**Comment by keita1523** [comment]
Github Actions are fail.

**Review by keita1523 — DISMISSED** [review]

File: CARET_trace/src/context.cpp (hunk 4 above)
Comment (hsgwa) [reply CARET_trace/src/context.cpp]: removed: [76d6293](https://github.com/tier4/CARET_trace/pull/68/commits/76d62938fb60a8a876270d924dc82c1395cb028f)

File: CARET_trace/src/ros_trace_points.cpp (hunk 15 above)
Comment (hsgwa) [reply CARET_trace/src/ros_trace_points.cpp]: addressed: [fa970d7](https://github.com/tier4/CARET_trace/pull/68/commits/fa970d7850562d102f303d87e68f9a4af81bb817)

File: CARET_trace/src/ros_trace_points.cpp (hunk 14 above)
Comment (hsgwa) [reply CARET_trace/src/ros_trace_points.cpp]: addressed: [fa970d7](https://github.com/tier4/CARET_trace/pull/68/commits/fa970d7850562d102f303d87e68f9a4af81bb817)

File: CARET_trace/src/ros_trace_points.cpp (hunk 12 above)
Comment (hsgwa) [reply CARET_trace/src/ros_trace_points.cpp]: addressed: [fa970d7](https://github.com/tier4/CARET_trace/pull/68/commits/fa970d7850562d102f303d87e68f9a4af81bb817)

File: CARET_trace/src/ros_trace_points.cpp (hunk 7 above)
Comment (hsgwa) [reply CARET_trace/src/ros_trace_points.cpp]: addressed: [fa970d7](https://github.com/tier4/CARET_trace/pull/68/commits/fa970d7850562d102f303d87e68f9a4af81bb817)

File: CARET_trace/src/ros_trace_points.cpp (hunk 8 above)
Comment (hsgwa) [reply CARET_trace/src/ros_trace_points.cpp]: addressed: [fa970d7](https://github.com/tier4/CARET_trace/pull/68/commits/fa970d7850562d102f303d87e68f9a4af81bb817)

File: CARET_trace/src/ros_trace_points.cpp (hunk 6 above)
Comment (hsgwa) [reply CARET_trace/src/ros_trace_points.cpp]: addressed: [fa970d7](https://github.com/tier4/CARET_trace/pull/68/commits/fa970d7850562d102f303d87e68f9a4af81bb817)

File: CARET_trace/src/ros_trace_points.cpp (hunk 9 above)
Comment (hsgwa) [reply CARET_trace/src/ros_trace_points.cpp]: addressed: [fa970d7](https://github.com/tier4/CARET_trace/pull/68/commits/fa970d7850562d102f303d87e68f9a4af81bb817)

File: CARET_trace/src/ros_trace_points.cpp (hunk 11 above)
Comment (hsgwa) [reply CARET_trace/src/ros_trace_points.cpp]: addressed: [fa970d7](https://github.com/tier4/CARET_trace/pull/68/commits/fa970d7850562d102f303d87e68f9a4af81bb817)

File: CARET_trace/include/caret_trace/recordable_data.hpp (hunk 3 above)
Comment (hsgwa) [reply CARET_trace/include/caret_trace/recordable_data.hpp]: addressed: [7f3f87e](https://github.com/tier4/CARET_trace/pull/68/commits/7f3f87ea264aab99006074c1bc55375264dc117e)

File: CARET_trace/src/lttng_session.cpp (hunk 5 above)
Comment (hsgwa) [reply CARET_trace/src/lttng_session.cpp]: addressed:

**Comment by hsgwa** [comment]
@keita1523 

Could you review again, please?
I'll ignore clang-tidy-differential for now.

**Review by keita1523 — DISMISSED** [review]
LGTM

**Review by keita1523 — APPROVED** [review]
LGTM

**Pull request merged**